src/main/java/org/sonar/plugins/groovy/codenarc/CodeNarcSourceAnalyzer.java

```diff
@@ -0,0 +1,102 @@
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02
+ */
+package org.sonar.plugins.groovy.codenarc;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.LinkedListMultimap;
+import com.google.common.collect.Maps;
+import com.google.common.collect.Multimap;
+import org.codenarc.analyzer.AbstractSourceAnalyzer;
+import org.codenarc.results.DirectoryResults;
+import org.codenarc.results.FileResults;
+import org.codenarc.results.Results;
+import org.codenarc.rule.Violation;
+import org.codenarc.ruleset.RuleSet;
+import org.codenarc.source.SourceFile;
+
+import java.io.File;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+
+public class CodeNarcSourceAnalyzer extends AbstractSourceAnalyzer {
+
+  private final Map<File, List<Violation>> violationsByFile = Maps.newHashMap();
+  private final List<File> sourceFiles;
+  private final File baseDir;
+
+  public CodeNarcSourceAnalyzer(List<File> sourceFiles, File baseDir) {
+    this.sourceFiles = sourceFiles;
+    this.baseDir = baseDir;
+  }
+
+  @Override
+  public Results analyze(RuleSet ruleSet) {
+
+    Multimap<File, FileResults> resultsByFileByDirectory = processFiles(ruleSet);
+    DirectoryResults results = processDirectories(baseDir, resultsByFileByDirectory);
+
+    return results;
+  }
+
+  private Multimap<File, FileResults> processFiles(RuleSet ruleSet) {
+    Multimap<File, FileResults> results = LinkedListMultimap.create();
+    for (File file : sourceFiles) {
+      List<Violation> violations = collectViolations(new SourceFile(file), ruleSet);
+      violationsByFile.put(file, violations);
+      FileResults result = new FileResults(file.getAbsolutePath(), violations);
+      results.put(file.getParentFile(), result);
+    }
+    return results;
+  }
+
+  private static DirectoryResults processDirectories(File baseDir, Multimap<File, FileResults> resultsByFileByDirectory) {
+    Map<File, DirectoryResults> results = Maps.newHashMap();
+    results.put(baseDir, new DirectoryResults());
+
+    // prepopulate all the levels
+    for (File directory : resultsByFileByDirectory.keySet()) {
+      File parent = directory;
+      while (!results.containsKey(parent)) {
+        results.put(parent, new DirectoryResults());
+        parent = parent.getParentFile();
+      }
+    }
+
+    // add all the results of the files
+    for (File directory : resultsByFileByDirectory.keySet()) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Combine this loop with the one that starts on line 75. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS3047)